src/mlpack/core/metrics/bleu_score_impl.hpp

```diff
@@ -0,0 +1,157 @@
+/**
+ * @file core/metrics/bleu_score_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of BLEUScore class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_METRICS_BLEU_SCORE_IMPLHPP
+#define MLPACK_CORE_METRICS_BLEU_SCORE_IMPLHPP
+
+// In case it hasn't been included.
+#include "bleu_score.hpp"
+
+namespace mlpack {
+namespace metric {
+
+template <typename ElemType, typename PrecisionType>
+BLEU<ElemType, PrecisionType>::BLEU(const size_t maxOrder) :
+  maxOrder(maxOrder),
+  translationLength(0),
+  referenceLength(0)
+{
+  // Nothing to do here.
+}
+
+template <typename ElemType, typename PrecisionType>
+template <typename WordVector>
+std::map<WordVector, size_t> BLEU<ElemType, PrecisionType>::GetNGrams(
+  const WordVector& segment)
+{
+  std::map<WordVector, size_t> ngramsCount;
+  for (size_t order = 1; order < maxOrder + 1; ++order)
+  {
+    for (size_t i = 0; i < segment.size() - order + 1; ++i)
+    {
+      WordVector seq = WordVector(segment.begin() + i,
+                                  segment.begin() + i + order);
+      ngramsCount[seq]++;
+    }
+  }
+  return ngramsCount;
+}
+
+template <typename ElemType, typename PrecisionType>
+template <typename ReferenceCorpusType, typename TranslationCorpusType>
+ElemType BLEU<ElemType, PrecisionType>::Evaluate(
+  const ReferenceCorpusType& referenceCorpus,
+  const TranslationCorpusType& translationCorpus,
+  const bool smooth)
+{
+  typedef typename TranslationCorpusType::value_type WordVector;
```

Comment:
Do you mind to add some comments on some important lines?

Reply:
Sure, I will do that by tomorrow.

Reply:
@zoq: I've added a few comments as you suggested. Let me know if I need to add any further comments.

Reply:
Yes, that looks great, super helpful to understand the code.